src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -1809,6 +1836,9 @@ private int tryExecute(final Runnable runnable) {
                 // we failed; we have to drop the queue size back down again to compensate before we can retry
                 if (! NO_QUEUE_LIMIT) decreaseQueueSize();
                 if (UPDATE_STATISTICS) spinMisses.increment();
+                if (!TAIL_LOCK) {
+                    return EXE_RETRY;
```

Comment:
This case allows many `TaskNode` objects to be allocated for a single operation where previously the same instance was reused.

Reply:
I see that it could be a problem and I can remove it or...we can find a different way to pool it  for later reuse: probably the retry logic for this case should be handled direrctly inside `tryAllocate`. 
Anyway I see that this case should get the same treatment of the others to avoid busy spinning due to a failure under heavy contention.

Reply:
Or we can implement the version with `getAndSetTail` that cannot fail :)

Reply:
It's a bit odd that the lock-free path disables the tryExecute main loop in favor of an additional outer loop. I like the idea of moving the yield/retry logic into the tryExecute method.

Reply:
> Or we can implement the version with getAndSetTail that cannot fail :)

I'm curious what this looks like :-)

Reply:
Re 

> t's a bit odd that the lock-free path disables the tryExecute main loop in favor of an additional outer loop. I like the idea of moving the yield/retry logic into the tryExecute method.

There was a good reason for that one :) 
In my original idea (never shared before) I wanted to provide a similar method as public to let a use to decide what to do when the failure happen due to excessive contention: this is usefull on event-loop like applications ie 
- try to submit, getting rejected
- batch locally
- try to submit at the and of the loop or later
A simple way to allow a backpressure mechanism.

> I'm curious what this looks like :-)
I've no idea if this can be implemented with the logic of this executor, but the idea is deadly simple, see: https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscLinkedQueue.java#L68

And the performance advantage under contention is just shocking (not joking).

Reply:
Anyway, in the next days I will produce a version that handle internally to `tryExecute` the  retry logic and maybe in some follow-up PRs a whole new logic, if possible ;)

Reply:
I've already implemented that same version, but probably there is an addition trick I could implement to help under contention ;)

Reply:
There is maybe an additional thing that could be implemented, but I need to check on `perfasm` if it's relevant ie a `cachedTail` to let the `if (tailNext instanceof TaskNode)` move one only if `this.tail` is changed, saving from hammering it with `tail.getNext()` despite the new "relaxed" logic.

Reply:
If we need to chat again I'm available in or Red Hat chat during the week ;)

Reply:
I'm not officially involved with this project, I hope I haven't created the wrong impression, I use it at work and have an interest high performance web servers :-)

Reply:
Tbh me too: I work in the messaging team in Red Hat but I often send PRs re "high performance stuff" around for hobby